flutter/tonic/dart_isolate_reloader.cc

```diff
@@ -0,0 +1,327 @@
+    // Already have a sticky error. Just drop this result.
+    return;
+  }
+
+  // Drop the lock temporarily around the call to Finish because it may
+  // trigger a recursive call into the tag handler.
+  monitor_->Exit();
+  Dart_Handle error_or_null = load_result->Finish();
+  monitor_->Enter();
+
+  if (!Dart_IsNull(error_or_null)) {
+    // Set sticky error.
+    load_error_ = error_or_null;
+  }
+}
+
+void DartIsolateReloader::ProcessResultQueueLocked() {
+  while (load_results_.size() > 0) {
+    // Grab the first load result.
+    LoadResult* result = load_results_.front();
+    load_results_.pop();
+    HandleLoadResultLocked(result);
+    // We are finished with this result now.
+    delete result;
+  }
+}
+
+bool DartIsolateReloader::IsCompleteLocked() {
+  return (pending_requests_ == 0) && load_results_.empty();
+}
+
+bool DartIsolateReloader::BlockUntilComplete() {
+  MonitorLocker ml(monitor_);
+
+  while (true) {
+    ProcessResultQueueLocked();
+
+    if (IsCompleteLocked()) {
+      break;
+    }
+
+    // Wait to be notified about new I/O results.
+    ml.Wait();
+  }
+  return !Dart_IsNull(load_error_);
+}
+
+Dart_Handle DartIsolateReloader::HandleLibraryTag(Dart_LibraryTag tag,
+                                                  Dart_Handle library,
+                                                  Dart_Handle url) {
+  if (tag == Dart_kCanonicalizeUrl) {
+    // Pass through to actual tag handler.
+    return DartLibraryLoader::HandleLibraryTag(tag, library, url);
+  }
+  DartState* dart_state = DartState::Current();
+  DCHECK(dart_state);
+  DartIsolateReloader* isolate_reloader = dart_state->isolate_reloader();
+  // The first call into the tag handler ends up blocking the calling thread
+  // until the entire reload has completed. All other calls into the
+  // tag handler schedule requests and return immediately.
+  const bool blocking_call = (tag == Dart_kScriptTag);
+  if (!isolate_reloader) {
+    // The first call into this tag handler must be for the script.
+    DCHECK(tag == Dart_kScriptTag);
+    isolate_reloader = new DartIsolateReloader(
+        dart_state->library_loader().library_provider());
+    // Associate the reloader with the isolate.
+    dart_state->set_isolate_reloader(isolate_reloader);
+    // Switch the tag handler.
+    Dart_SetLibraryTagHandler(DartIsolateReloader::HandleLibraryTag);
+  } else {
+    // We should not see another request for the script.
+    DCHECK(tag != Dart_kScriptTag);
+  }
+
+  // Issue I/O request.
+  isolate_reloader->SendRequest(tag,
+                                url,
+                                (library != Dart_Null()) ?
+                                Dart_LibraryUrl(library) : Dart_Null());
+
+  if (blocking_call) {
+    // Block and process LoadResults until the load is complete.
+    const bool load_error = isolate_reloader->BlockUntilComplete();
+    // Disassociate this reloader from the isolate.
+    dart_state->set_isolate_reloader(NULL);
```

Comment:
`nullptr`

We should take the `std::unique_ptr` into a local variable so that we free it using the `std::unique_ptr` destructor